python/source/impedance_spectroscopy.py

```diff
@@ -188,3 +199,184 @@ def fourier_analysis(data, ptree=None):
     impedance = fft_voltage[peak_indices]/fft_current[peak_indices]
 
     return [frequency, impedance]
+
+class NyquistPlot(Observer):
+    '''Nyquist plot.
+
+    The frequency-dependent impedance Z(f) is represented as a complex number:
+        Z(f) = R + jX
+    In a Nyquist plot, the real part of the impedance R = real(Z) is plotted on
+    the X-axis and the imaginary part X = imag(Z) on the Y-axis. R and X are
+    called resistance and reactance, respectively. Both are expressed in ohm.
+
+    Nyquist plots have one major shortcoming. When looking at any data point
+    on the complex plane, it is impossible to tell what frequency was used to
+    record that point.
+
+    See also
+    --------
+    BodePlot
+    '''
+    def __new__(cls, *args, **kwargs):
+        return object.__new__(NyquistPlot)
+    def __init__(self, filename=None):
+        self._figure = pyplot.figure(figsize=(14, 14))
+        self._filename = filename
+    def update(self, subject, *args, **kwargs):
+        plot_nyquist(subject._data, figure=self._figure)
+        if self._filename is not None:
+            pyplot.savefig(self._filename, bbox_inches='tight')
+Observer._builders['NyquistPlot'] = NyquistPlot
+
+class BodePlot(Observer):
+    '''Bode plot.
+
+    A Bode plot is another popular presentation method for impedance spectra.
+    The impedance is plotted with frequency on the X-axis in log scale and
+    absolute values of the impedance in dB, 20*log10(|Z|), and the phase shift
```

Comment:
yes certain.
"Since the decibel is defined with respect to power, not amplitude, conversions of voltage ratios to decibels must square the amplitude, or use the factor of 20 instead of 10, as discussed above." [From Wikipedia https://en.wikipedia.org/wiki/Decibel]